stored_requests/config/config.go

```diff
@@ -34,15 +34,20 @@ import (
 //
 // As a side-effect, it will add some endpoints to the router if the config calls for it.
 // In the future we should look for ways to simplify this so that it's not doing two things.
-func NewStoredRequests(cfg *config.StoredRequests, client *http.Client, router *httprouter.Router) (fetcher stored_requests.Fetcher, ampFetcher stored_requests.Fetcher, db *sql.DB, shutdown func()) {
-	if cfg.Postgres.ConnectionInfo.Database != "" {
-		glog.Infof("Connecting to Postgres for Stored Requests. DB=%s, host=%s, port=%d, user=%s", cfg.Postgres.ConnectionInfo.Database, cfg.Postgres.ConnectionInfo.Host, cfg.Postgres.ConnectionInfo.Port, cfg.Postgres.ConnectionInfo.Username)
-		db = newPostgresDB(cfg.Postgres.ConnectionInfo)
-	}
-	eventProducers, ampEventProducers := newEventProducers(cfg, client, db, router)
-	cache := newCache(cfg)
-	ampCache := newCache(cfg)
-	fetcher, ampFetcher = newFetchers(cfg, client, db)
+func NewStoredRequests(cfg *config.Configuration, client *http.Client, router *httprouter.Router) (fetcher stored_requests.Fetcher, ampFetcher stored_requests.Fetcher, db *sql.DB, shutdown func(), categoriesFetcher stored_requests.Fetcher) {
+	if cfg.StoredRequests.Postgres.ConnectionInfo.Database != "" {
+		glog.Infof("Connecting to Postgres for Stored Requests. DB=%s, host=%s, port=%d, user=%s",
+			cfg.StoredRequests.Postgres.ConnectionInfo.Database,
+			cfg.StoredRequests.Postgres.ConnectionInfo.Host,
+			cfg.StoredRequests.Postgres.ConnectionInfo.Port,
+			cfg.StoredRequests.Postgres.ConnectionInfo.Username)
+		db = newPostgresDB(cfg.StoredRequests.Postgres.ConnectionInfo)
+	}
+	eventProducers, ampEventProducers := newEventProducers(&cfg.StoredRequests, client, db, router)
+	cache := newCache(&cfg.StoredRequests)
+	ampCache := newCache(&cfg.StoredRequests)
+	fetcher, ampFetcher = newFetchers(&cfg.StoredRequests, client, db)
+	categoriesFetcher, _ = newFetchers(&cfg.CategoryMapping, client, db)
```

Comment:
I am thinking we should cast this to a categories fetcher at this point, so there is no confusion that this is a category fetcher, not a standard fetcher.

Actually, to be secure, I think we should define a `globalFetcher` interface that implements both fetcher and categoryFetcher, and have that be what `newFetchers()` returns. The makes explicit that `newFetchers()` supports both, and then we immediately cast them into the fetchers we want at this point. This would also force us to make fetcher decorators like cache support need to explicitly support category fetching from the get go, but otherwise we can run into the case where caching only supports the plain fetcher, so when it is applied, the result cannot be cast to a category fetcher, and you have a crash based on a seemingly valid config not working as intended.

Reply:
To be honest, at this point I would say we want to have `newFetchers()` to return a standard, amp, or category fetcher, and have us call it three times. But that would require refactoring a significant chunk of code, and probably changing the layout of the config file, so out of scope for this PR.

Reply:
That's a good call out. I like your idea of having newFetcher() that returns a single instance and we call it three times, but agreed that it will affect a lot of areas of the code so probably best done in a separate PR to avoid bloating this one.

Reply:
There is still the suggestion to have `newFetchers()` return an interface that supports both methods, and then immediately cast them to a normal fetcher interface or a category fetcher interface.